src/tribol/search/InterfacePairFinder.cpp

```diff
@@ -1,643 +1,873 @@
+  {
+    constexpr double sc = 1./3.;
 
-private:
-   /*!
-    * Expands bounding box by 33% of longest dimension's range
-    */
-   void inflateBBox(SpatialBoundingBox& bbox)
-   {
-      constexpr double sc = 1./3.;
+    int d = bbox.getLongestDimension();
+    const RealT expansionFac =  sc * bbox.range()[d];
+    bbox.expand(expansionFac);
+  }
 
-      int d = bbox.getLongestDimension();
-      const real expansionFac =  sc * bbox.range()[d];
-      bbox.expand(expansionFac);
-   }
+  CouplingScheme* m_coupling_scheme;
+  const MeshData::Viewer& m_mesh1;
+  const MeshData::Viewer& m_mesh2;
 
-private:
+  ImplicitGridType m_grid;
+  SpatialBoundingBox m_gridBBox;
+  ArrayT<SpatialBoundingBox> m_meshBBoxes1;
 
-   CouplingScheme* m_couplingScheme;
-   MeshWrapper<D> m_meshWrapper1;
-   MeshWrapper<D> m_meshWrapper2;
+}; // End of GridSearch class definition
 
-   ImplicitGridType m_grid;
-   SpatialBoundingBox m_gridBBox;
-   containerArray<SpatialBoundingBox> m_meshBBoxes1;
 
-};
+///////////////////////////////////////////////////////////////////////////////
 
 /*!
- * Compute all pairs of elements in the two meshes of the CouplingScheme
+ * \brief BVH helper class to compute the candidate pairs for a coupling scheme
  *
- * \note Assumes the two meshes are different
+ * A BvhSearch indexes the elements from the first mesh of
+ * the coupling scheme in a spatial index that requires element bounding boxes.
+ * Then, for each of the elements in the second mesh, we find the candidate
+ * pairs and add them to the coupling scheme's list of contact pairs.
+ *
+ * The spatial index is generated in \a generateSpatialIndex()
+ * and the search is performed in \a findInterfacePairs()
+ * 
+ * \tparam D The spatial dimension of the coupling scheme mesh vertices.
  */
- void generateCartesianProductPairs(CouplingScheme* cs)
- {
-    MeshManager & meshManager = MeshManager::getInstance();
-
-    integer meshId1 = cs->getMeshId1();
-    MeshData const & meshData1 = meshManager.GetMeshInstance(meshId1);
-    integer mesh1NumElems = meshData1.m_numCells;
-
-    integer meshId2 = cs->getMeshId2();
-    MeshData const & meshData2 = meshManager.GetMeshInstance(meshId2);
-    integer mesh2NumElems = meshData2.m_numCells;
-
-    int numPairs = mesh1NumElems * mesh2NumElems;
-
-    InterfacePairs* contactPairs = cs->getInterfacePairs();
-    contactPairs->clear();
-    contactPairs->reserve( numPairs );
+template<int D, class ExecSpace>
+class BvhSearch : public SearchBase
+{
+public:
+  using BVHT = axom::spin::BVH<D, ExecSpace, RealT>;
+  using BoxT = typename BVHT::BoxType;
+  using PointT = primal::Point<RealT, D>;
+  using RayT = primal::Ray<RealT, D>;
+  using VectorT = primal::Vector<RealT, D>;
+  using AtomicPolicy = typename axom::execution_space<ExecSpace>::atomic_policy;
+
+  /*!
+  * Constructs a BvhSearch instance over CouplingScheme \a couplingScheme
+  * \pre couplingScheme is not null
+  */
+  BvhSearch(CouplingScheme* coupling_scheme)
```

Comment:
Has the BVH search been tested against, let's say, cartesian product to make sure we get what we are expecting?

Reply:
Check for existing tests; add BVH to them